### ISSUE TITLE
Lint/Void autocorrection is unsafe when a constant is referenced

### ISSUE THREAD

**Issue by jonleighton** [rubocop/rubocop#12017]
When I run `rubocop -a` on this file:

``` ruby
Foo::Bar

class Foo::Bar
end
```

It rewrites the file like so:

``` ruby
class Foo::Bar
end
```

However, this is unsafe, due to Ruby's constant autoloading mechanism.

For example, in a Rails app, the constant may be referenced to force it to be loaded from another file, before apply a monkey-patch.

--------

## Expected behavior

Either the file is unchanged, or the change is only applied when unsafe autocorrections are used.

## Actual behavior

See above.

## Steps to reproduce the problem

1. Put the above into `test.rb`
2. Run `rubocop -a test.rb`

## RuboCop version

```
1.54.0 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [aarch64-linux-musl]
```

**Comment by bbatsov** [comment]
Reopened — still reproduces on master (1.88): `rubocop -a` (safe autocorrect) removes a load-bearing bare `Foo::Bar` constant reference. It should be left alone, or only corrected under `-A`.
